### PR TITLE
add(README.md): Documentation for the env-injector snapcraft extension

### PR DESCRIPTION
**Pull request by locnnil**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
+# Environment Variables Injector
```
Comment (farshidtz) [inline README.md]:
```suggestion
# The env-injector extension
```

Consistent with other extension docs: https://snapcraft.io/docs/supported-extensions

File: README.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+This extension is designed to help users export snap options configured with the [snap set command](https://snapcraft.io/docs/configuration-in-snaps) into environment variables.
```
Comment (farshidtz) [inline README.md]: There are a few ways to set configurations, not just via the `snap set` command.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+Additionally, this extension supports environment files, which we refer to as `envfile` for brevity.
```
Comment (farshidtz) [inline README.md]: The introduction needs improvements, keeping the following points in mind:
- The extension is to inject environment variables into snaps
- The extension offers two ways to do it
  - Via snap configuration options
  - Via environment files or env files

File: README.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+| Global | env.<key>=<value>              | <KEY>=<value>        |
+| Local  | apps.<app>.env.<key>=<value>   | <KEY>=<value>        |
+
+**Examples**
+
+Assuming `my-snap` is the name of the snap that uses the env-injector extension:
+
+* `sudo snap set <my-snap> env.default_port=8080` will be mapped to `DEFAULT_PORT=8080` globally.
+
+* `sudo snap set <my-snap> apps.server.env.endpoint="/v2/store"` will be mapped to `ENDPOINT="/v2/store"` and will be visible only to the `server` app.
```
Comment (farshidtz) [inline README.md]: Turn into a subsection, followed by another one for environment files

File: README.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+For environment files, where `path` refers to the `envfile` path:
```
Comment (farshidtz) [inline README.md]: `envfile` is not an alias for environment files. It is part of the snap option key for environment file / env file path.

`envfile` - > env file

File: README.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+|        | Snap Configuration Option    |
+|--------|------------------------------|
+| Global | envfile=<path>               |
+| Local  | apps.<app>.envfile=<path>    |
+
+**Examples**
+
+* `sudo snap set <my-snap> envfile=$HOME/envfile.env` makes all the environment variables globally visible.
+
+* `sudo snap set <my-snap> apps.server.envfile=$HOME/envfile.env` makes the environment variables visible only to the `server` app.
```
Comment (farshidtz) [inline README.md]: File access requirements not explained.

File: README.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+## Syntax and Rules
+
+The precedence of environment variables follows this order:
+
+1. Env vars from local snap config options.
+2. Env vars from global snap config options.
+3. Env vars extracted from the **local** `envfile`.
+4. Env vars extracted from the **global** `envfile`.
```
Comment (farshidtz) [inline README.md]: Reading this sounds like the global env file has the highest precedence, applied in the end.

Comment (farshidtz) [inline README.md]:
```suggestion
3. Env vars extracted from the **local** env file.
4. Env vars extracted from the **global** env file.
```

File: README.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+|        | Snap Configuration Option      | Environment Variable |
+|--------|--------------------------------|----------------------|
+| Global | env.<key>=<value>              | <KEY>=<value>        |
+| Local  | apps.<app>.env.<key>=<value>   | <KEY>=<value>        |
+
+**Examples**
+
+Assuming `my-snap` is the name of the snap that uses the env-injector extension:
+
+* `sudo snap set <my-snap> env.default_port=8080` will be mapped to `DEFAULT_PORT=8080` globally.
```
Comment (farshidtz) [inline README.md]:
```suggestion
* `sudo snap set <my-snap> env.default-port=8080` will be mapped to `DEFAULT_PORT=8080` and exported globally.
```

File: README.md (hunk 4 above)
Comment (farshidtz) [inline README.md]:
```suggestion
* `sudo snap set <my-snap> apps.server.env.endpoint="/v2/store"` will be mapped to `ENDPOINT="/v2/store"` and exported only to the `server` app.
```

File: README.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+3. Env vars extracted from the **local** `envfile`.
+4. Env vars extracted from the **global** `envfile`.
+
+| Accepted Characters in Snap Option Keys                   | Mapped Environment Variable Names |
+|-----------------------------------------------------------|-----------------------------------|
+| Lowercase letters                                         | Uppercase letters                 |
+| Numbers (not at the beginning and with lowercase letters) | Numbers                           |
+| Hyphens (surrounded by lowercase letters)                 | Underscores                       |
+
+Any characters not listed in the first column of the above table are not allowed.
```
Comment (farshidtz) [inline README.md]: Move to snap options subsection.

File: README.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+- It makes a request to the [snapd API](https://snapcraft.io/docs/using-the-api) through the snapd Unix socket.
+
+- It reads the available environment variables or paths to the envfiles.
+
+- It translates the snap options into environment variables.
+
+- If available, it sources the envfiles.
+
+- It sources all envfiles.
```
Comment (farshidtz) [inline README.md]:
```suggestion
- It reads the available environment variables or paths to the env files.

- It translates the snap options into environment variables.

- If available, it sources the env files.

- It sources all env files.
```

File: README.md [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+Any characters not listed in the first column of the above table are not allowed.
+
+## Exporter Program Workflow
+
+The process works as follows:
+
+### During Build:
+
+- An application responsible for dynamically exporting environment variables is added to the snap's [command chain](https://snapcraft.io/docs/snapcraft-app-and-service-metadata#heading--command-chain).
```
Comment (farshidtz) [inline README.md]:
```suggestion
- An application responsible for dynamically exporting environment variables is added to the needed apps's [command chain](https://snapcraft.io/docs/snapcraft-app-and-service-metadata#heading--command-chain).
```

File: README.md (hunk 10 above)
Comment (farshidtz) [inline README.md]: Overlapping logic.

File: README.md [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+- It reads the available environment variables or paths to the envfiles.
+
+- It translates the snap options into environment variables.
+
+- If available, it sources the envfiles.
+
+- It sources all envfiles.
+
+- It sources all variables.
```
Comment (farshidtz) [inline README.md]: Not sure what this means.

File: README.md
Comment (farshidtz) [inline README.md:1]: I'd add a section on Using the extension and give a snapcraft yaml example with a couple of apps using the extension. Then move the usage snap option examples to there. Also, document the alias.

**Review by farshidtz — CHANGES_REQUESTED** [review]
This is a great start. Please see my comments and also refer to the spec and borrow as much text as you can.

Thanks

File: README.md (hunk 7 above)
Comment (locnnil) [reply README.md]: Actually, this is the order of precedence that we use, but we can also list by **the order which is processed/applied**, which will be exactly the inverse of this list.

File: README.md (hunk 1 above)
Comment (locnnil) [reply README.md]: Yeah, that makes sense! Thanks!

File: README.md (hunk 12 above)
Comment (locnnil) [reply README.md]: First it processes the env files, then it processes the snap configs.

File: README.md (hunk 7 above)
Comment (farshidtz) [reply README.md]: You could change the first sentence to "The environment variables are processed in the following order:" to avoid ambiguity.

File: README.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+Additionally, this extension supports environment files, which we refer to as `envfile` for brevity.
+
+To prevent conflicts with other snap configurations and to maintain some level of isolation between apps, the following rules apply:
+
+* **Global:** Environment variables are visible to all apps within the snap that use this extension.
+* **Local:** Environment variables are visible only to the specific `app` defined during the snap set command.
+
+|        | Snap Configuration Option      | Environment Variable |
+|--------|--------------------------------|----------------------|
+| Global | env.<key>=<value>              | <KEY>=<value>        |
```
Comment (farshidtz) [inline README.md]: Avoid HTML syntax as they don't get rendered as intended.
```suggestion
| Global | `env.<key>=<value>`              | `<KEY>=<value>`        |
```

Comment (locnnil) [reply README.md]: Fixed! Thanks!

File: README.md (hunk 7 above)
Comment (locnnil) [reply README.md]: Done! Thank you!

File: README.md (hunk 2 above)
Comment (locnnil) [reply README.md]: Addressed, thank you!

File: README.md (hunk 3 above)
Comment (locnnil) [reply README.md]: Addressed on https://github.com/locnnil/env-inject-doc/pull/1/commits/1f0181a11f9680f197444887ec495a6854d5e13b! Thanks!

File: README.md (hunk 4 above)
Comment (locnnil) [reply README.md]: Done in https://github.com/locnnil/env-inject-doc/pull/1/commits/1f0181a11f9680f197444887ec495a6854d5e13b.

File: README.md (hunk 5 above)
Comment (locnnil) [reply README.md]: Sure, fixed! On 033f468

File: README.md (hunk 6 above)
Comment (locnnil) [reply README.md]: Explained, thanks!

File: README.md (hunk 9 above)
Comment (locnnil) [reply README.md]: Moved! Thank you!

File: README.md (hunk 10 above)
Comment (locnnil) [reply README.md]: Fixed! Thank you!

File: README.md
Comment (locnnil) [reply README.md:1]: Added a "**How to use it**" section, as on the others extensions.

File: README.md [hunk 14]
```diff
@@ -0,0 +1,144 @@
+# The env-injector extension
+
+The env-injector extension allows developers to export snap options as environment variables within the snap sandbox.
+These options can be configured using the [snap set command](https://snapcraft.io/docs/configuration-in-snaps) or via the [the snap API](https://snapcraft.io/docs/using-the-api).
+This extension dynamically converts snap configuration options into environment variables, which are then accessible to the app within the snap sandbox.
+For this to work, the snap options must follow a specific pattern that the extension can match.
+The extension also supports environment files, where the snap configuration options specify the path to the environment file.
```
Comment (farshidtz) [inline README.md]: The first sentence isn't accurate. I'd rewrite the whole paragraph with both functions in mind (snap options and env files):

```suggestion
The env-injector extension adds support for passing user-defined environment variables to snaps. 
By adding the extension to snapped apps, the user of the snap will be allowed to set environment variables via [snap configuration options](https://snapcraft.io/docs/configuration-in-snaps) or environment files and have them exported to the apps on startup.
```

File: README.md [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+
+
+## How to use it
+
+Add `extensions: [ env-injector ]` to the application definition in your `snapcraft.yaml` file.
+By default the extension defines an app alias in the format of an env variable to be used in the snap config with the same name as defined in your apps section of the `yaml`.
+But if you wish, you can avoid to use this default by creating an environment Variable yourself with named as `env_alias` with the name that you wish to use to set the environment variables to your app.
```
Comment (farshidtz) [inline README.md]:
```suggestion
```

File: README.md [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+
+Add `extensions: [ env-injector ]` to the application definition in your `snapcraft.yaml` file.
+
+By default, this extension creates an environment variable with the same name as the app defined in the yaml, which can be used in the snap configuration.
+However, if you prefer, you can define this environment variable yourself by using `env_alias` with the name you want to use for setting the environment variables in your app.
+
+- Default usage example:
```
Comment (farshidtz) [inline README.md]:
```suggestion
Example, adding the extension to an app named `rabbitmq`:
```

File: README.md [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+```yaml
+apps:
+  rabbitmq:
+    command: bin/run.sh
+    daemon: simple
+    extensions: [ env-injector ]
+```
+
+This way you can define environment variables to your application by doing:
```
Comment (farshidtz) [inline README.md]:
```suggestion
Then, the user of the snap can pass environment variables via snap options or env files.

For example, to export `RABBITMQ_DEFAULT_PASS=password` using the `snap set` command:
```

File: README.md [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+| Global | `envfile=<path>`            |
+| Local  | `apps.<app>.envfile=<path>` |
+
+**Examples**
+
+* `sudo snap set <my-snap> envfile=$HOME/envfile.env` makes all the environment variables globally visible.
+
+* `sudo snap set <my-snap> apps.server.envfile=$HOME/envfile.env` makes the environment variables visible only to the `server` app.
+
+> [!NOTE]
```
Comment (farshidtz) [inline README.md]: This is Github-specific syntax.

**Review by farshidtz — CHANGES_REQUESTED** [review]
Thanks for the changes. I find this hard to follow, especially because env files are treated as an additional thing rather than a first party feature. The syntax and rules are scattered all over the text. I will propose some changes via a PR.

PR: https://github.com/locnnil/env-inject-doc/pull/2

**Comment by locnnil** [comment]
At https://github.com/locnnil/env-inject-doc/pull/1/commits/55a87ea47cc8f9db8d7720bb039c4f80b397b44a I've just reordered the tables so the syntax for the snap options and the env file stay together and then the table with the allowed characters show up. 

Also, I've added a tiny introduction sentence [here](https://github.com/locnnil/env-inject-doc/blob/55a87ea47cc8f9db8d7720bb039c4f80b397b44a/README.md?plain=1#L146) because an `H2 title` followed directly by a `H3 title` without nothing in the middle seems ugly.

File: README.md [hunk 19]
```diff
@@ -106,6 +100,14 @@ Snap configuration options for setting env file paths:
...
 
 
+| Accepted characters in snap option keys                   | Mapped environment variable names |
+|-----------------------------------------------------------|-----------------------------------|
+| Lowercase letters                                         | Uppercase letters                 |
+| Numbers (not at the beginning and with lowercase letters) | Numbers                           |
+| Hyphens (surrounded by lowercase letters)                 | Underscores                       |
+
+Any characters not listed in the first column of the above table are not allowed.
+
```
Comment (farshidtz) [inline README.md:102]: Two issues with https://github.com/locnnil/env-inject-doc/pull/1/commits/55a87ea47cc8f9db8d7720bb039c4f80b397b44a:
- Line 93 which introduces the table hasn't been moved.
- This move results in confusion. The character table is unrelated to env file options but rather there to explain the mapping between snap options and environment variable names.

**Review by farshidtz — CHANGES_REQUESTED** [review]

File: README.md (hunk 19 above)
Comment (locnnil) [reply README.md:102]: True! Fixed!  Thanks!

File: README.md [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+| Local  | `apps.<app>.envfile=<path>` |
+
+### Order precedence
+
+The environment variables are processed in the following order:
+
+1. Env vars extracted from the **global** env file.
+2. Env vars extracted from the **local** env file.
+3. Env vars from global snap config options.
+4. Env vars from local snap config options.
```
Comment (farshidtz) [inline README.md]:
```suggestion
1. Env vars extracted from the global env file.
2. Env vars extracted from the local env file.
3. Env vars from global snap config options.
4. Env vars from local snap config options.
```

**Review by farshidtz — APPROVED** [review]
Looks good, thanks.

File: README.md [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+Example, adding the extension to an app named `server`:
+```yaml
+apps:
+  server:
+    command: run.sh
+    daemon: simple
+    extensions: [ env-injector ]
+```
+
+With this, the user of the snap will be enabled to pass environment variables via snap options or env files.
```
Comment (jpm-canonical) [inline README.md:19]: s/enabled/able

File: README.md [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+
+```bash
+sudo snap set <snap-name> env.http-port=8080
+```
+
+Internally, the extension converts `http-port` to `HTTP_PORT` and exports it before executing the app's command.
+
+The above syntax is to set environment variables for all the apps that use this extension within the snap.
+We refer to these as *global options*. 
+With the given example, the snap only has one app which uses the extension so setting a global option is acceptable.
```
Comment (jpm-canonical) [inline README.md:33]: `extension, `

File: README.md [hunk 23]
```diff
@@ -0,0 +1,171 @@
...
+
+Refer [here](#syntax) for more details on the syntax.
+
+
+### Pass environment variables via env files
+
+A list of environment variables can be passed to the snap via one or more environment files.
+
+> **Note**  
+> For confined snaps, the files need to be accessible from within the snap, either by placing them in the [writable area](https://snapcraft.io/docs/data-locations) of the snap or via a file access interface, such as the [home](https://snapcraft.io/docs/home-interface) or [personal-files](https://snapcraft.io/docs/personal-files-interface).
```
Comment (jpm-canonical) [inline README.md:54]: Either drop `the` before `[home]`, or add `interface` at the end.

File: README.md [hunk 24]
```diff
@@ -0,0 +1,171 @@
...
+* **Local:** Environment variables are visible only to the specific `app` defined during the snap set command.
+
+Snap configuration options to environment variable mapping:
+
+|        | Snap configuration option      | Environment variable |
+|--------|--------------------------------|----------------------|
+| Global | `env.<key>=<value>`            | `<KEY>=<value>`      |
+| Local  | `apps.<app>.env.<key>=<value>` | `<KEY>=<value>`      |
+
+The `key` is converted to environment variable name based on the following rules:
```
Comment (jpm-canonical) [inline README.md:93]: `...to an environment..`

File: README.md [hunk 25]
```diff
@@ -0,0 +1,171 @@
...
+
+| Accepted characters in snap option keys                   | Mapped environment variable names |
+|-----------------------------------------------------------|-----------------------------------|
+| Lowercase letters                                         | Uppercase letters                 |
+| Numbers (not at the beginning and with lowercase letters) | Numbers                           |
```
Comment (jpm-canonical) [inline README.md:98]: I don't understand why `and with lowercase letters` is mentioned here

File: README.md [hunk 26]
```diff
@@ -0,0 +1,171 @@
...
+| Hyphens (surrounded by lowercase letters)                 | Underscores                       |
```
Comment (jpm-canonical) [inline README.md:99]: I don't think it's necessary to add the phrase in brackets. It should be clear enough from the first line in this table.

File: README.md [hunk 27]
```diff
@@ -0,0 +1,171 @@
...
+
+Any characters not listed in the first column of the above table are not allowed.
```
Comment (jpm-canonical) [inline README.md:101]: `the above table` => `the table above`

or just say:
`Any characters not listed in this table are not allowed.`

**Review by jpm-canonical — COMMENTED** [review]
A few small comments

File: README.md (hunk 21 above)
Comment (locnnil) [reply README.md:19]: Done! Thanks!
